score.go

```diff
@@ -0,0 +1,952 @@
+			if p1 > topicParams.TimeInMeshCap {
+				p1 = topicParams.TimeInMeshCap
+			}
+			topicScore += p1 * topicParams.TimeInMeshWeight
+		}
+
+		// P2: first message deliveries
+		p2 := tstats.firstMessageDeliveries
+		topicScore += p2 * topicParams.FirstMessageDeliveriesWeight
+
+		// P3: mesh message deliveries
+		if tstats.meshMessageDeliveriesActive {
+			if tstats.meshMessageDeliveries < topicParams.MeshMessageDeliveriesThreshold {
+				deficit := topicParams.MeshMessageDeliveriesThreshold - tstats.meshMessageDeliveries
+				p3 := deficit * deficit
+				topicScore += p3 * topicParams.MeshMessageDeliveriesWeight
+			}
+		}
+
+		// P3b:
+		p3b := tstats.meshFailurePenalty
+		topicScore += p3b * topicParams.MeshFailurePenaltyWeight
+
+		// P4: invalid messages
+		p4 := tstats.invalidMessageDeliveries
+		topicScore += p4 * topicParams.InvalidMessageDeliveriesWeight
+
+		// update score, mixing with topic weight
+		score += topicScore * topicParams.TopicWeight
+	}
+
+	// P5: application-specific score
+	p5 := ps.params.AppSpecificScore(p)
+	score += p5 * ps.params.AppSpecificWeight
+
+	// P6: IP collocation factor
+	for _, ip := range pstats.ips {
+		peersInIP := len(ps.peerIPs[ip])
+		if peersInIP > ps.params.IPColocationFactorThreshold {
+			surpluss := float64(peersInIP - ps.params.IPColocationFactorThreshold)
+			p6 := surpluss * surpluss
+			score += p6 * ps.params.IPColocationFactorWeight
+		}
+	}
+
+	return score
+}
+
+// periodic maintenance
+func (ps *peerScore) background(ctx context.Context) {
+	refreshScores := time.NewTicker(ps.params.DecayInterval)
+	defer refreshScores.Stop()
+
+	refreshIPs := time.NewTicker(time.Minute)
+	defer refreshIPs.Stop()
+
+	gcDeliveryRecords := time.NewTicker(time.Minute)
+	defer gcDeliveryRecords.Stop()
+
+	var inspectScores <-chan time.Time
+	if ps.inspect != nil {
+		ticker := time.NewTicker(ps.inspectPeriod)
+		defer ticker.Stop()
+		inspectScores = ticker.C
+	} else {
+		// never fires
+		inspectScores = make(chan time.Time)
+	}
+
+	for {
+		select {
+		case <-refreshScores.C:
+			ps.refreshScores()
+
+		case <-refreshIPs.C:
+			ps.refreshIPs()
+
+		case <-gcDeliveryRecords.C:
+			ps.gcDeliveryRecords()
+
+		case <-inspectScores:
+			ps.inspectScores()
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (ps *peerScore) inspectScores() {
+	ps.Lock()
+	scores := make(map[peer.ID]float64, len(ps.peerStats))
+	for p := range ps.peerStats {
+		scores[p] = ps.score(p)
+	}
+	ps.Unlock()
+
+	ps.inspect(scores)
+}
+
+func (ps *peerScore) refreshScores() {
+	ps.Lock()
+	defer ps.Unlock()
+
+	now := time.Now()
+	for p, pstats := range ps.peerStats {
+		if !pstats.connected {
+			// has the retention period expired?
+			if now.After(pstats.expire) {
+				// yes, throw it away (but clean up the IP tracking first)
+				ps.removeIPs(p, pstats.ips)
+				delete(ps.peerStats, p)
+			}
+
+			// we don't decay retained scores, as the peer is not active.
```

Comment:
Can a peer disconnect and connect seconds before the retention period elapses, in order to keep their score "alive" perpetually? We may want to introduce additional measures here, e.g.:

1. Tallying periods of connectedness/disconnectedness across interruptions. When the total disconnectedness time surprasses a threshold, we forget the peer's score.
2. Tracking absolute number of disconnections and penalising for interruptions, relative to interruptions from other peers => rationale: if we have a high interruption rate from all peers, we might be the ones that are unstable and don't want to penalise anybody because we'll be shooting ourselves in the foot.

Reply:
We could, but that just adds extra complexity here. Can we punt on this until we have a clear idea that it is a problem and how to resolve it?

Reply:
Complexity how? Counting connections and disconnections is simple; aren't we already subscribing to connectedness events?

Anyway, I think it's trivial to game the scoring system right now, and it shouldn't be this simple. cc @yiannisbot @daviddias

Reply:
That's a heavy statement. How would you game the score by connecting and disconnecting?

Reply:
we can make it so that only negative scores are retained if you are so concerned about this.

Reply:
> How would you game the score by connecting and disconnecting?

1. Attacker gets GRAFT'ed onto a topic.
2. Attacker behaves well, accumulates positive score.
3. Attacker disconnects. Their positive score is retained for `RetainScore` time.
4. Just before `RetainScore` elapses, the attacker connects again. This refreshes their expiry time.
5. They wait a little, so that a score recalculating kicks in. This line will actually award them points for the time they've been disconnected: https://github.com/libp2p/go-libp2p-pubsub/blob/feat/hardening/score.go#L517 
6. They disconnect.
7. Lather, rinse, repeat. You keep building up your P1 score without even being connected.

Reply:
Ok, that's a valid concern. The simple (and correct) solution is to only retain negative scores, so that the attacker can't build up score by using this method.

Reply:
P1 is only valid while they are in the mesh, it is not retained btw.

Reply:
So your attack is not plausible against P1, this simply disappears when they leave the mesh.
It is however a valid concern about building score by publishing messages with the first message deliveries counter.

Reply:
@vyzo 

> P1 is only valid while they are in the mesh, it is not retained btw.

👍

> The simple (and correct) solution is to only retain negative scores, so that the attacker can't build up score by using this method.

I think we need something more sophisticated, though. I wouldn't like a peer to lose all their positive karma due to a 1-sec hiccup. That's why I suggested some form of failure counting and tolerance, or live/alive tallying, after which, yes, we'd reset any positive scores.

Reply:
Following sync discussion, it's hard to build logic around disconnections because we don't know the reason of the disconnection -- it could just be the connection manager kicking in and us killing the connection.

Reply:
I implemented negative score retention; take another look.
When we have a better idea how to retain positive scores, we can add the relevant logic.